gitlabform/processors/group/group_members_processor.py

```diff
@@ -229,8 +247,48 @@ def _process_users(
                     )
                     continue
                 debug("Removing user '%s' who is not configured to be a member.", user)
-                self.gitlab.remove_member_from_group(group, user)
+                user_id = self.get_user_id(user)
+                try:
+                    group.members.delete(user_id)
+                except GitlabDeleteError:
+                    pass
         else:
             debug("Not enforcing group members.")
 
-        debug("Group members AFTER: %s", self.gitlab.get_group_members(group))
+        debug("Group members AFTER: %s", group.members.list())
+
+    @staticmethod
+    def get_group_members(group) -> dict:
+        members = group.members.list()
+        users_before = {}
+        for member in members:
+            users_before[member.username.lower()] = member
+        return users_before
+
+    def get_user_id(self, username) -> str:
+        user = self._get_user_by_username(username)
+        return user.id
+
+    def get_group_id(self, groupname) -> int:
+        groups = self._get_group_by_groupname(groupname)
+        return groups.id
+
+    def get_group_by_name(self, groupname) -> Group:
+        group = self._get_group_by_groupname(groupname)
+        return self.gl.groups.get(group.id)
+
+    def _get_group_by_groupname(self, groupname):
+        groups = self.gl.groups.list(search=groupname)
+        if len(groups) == 0:
+            raise GitlabGetError(
+                "No groups found when searching for groupname '%s'" % groupname, 404
+            )
+        return groups[0]
+
+    def _get_user_by_username(self, username):
+        user = self.gl.users.list(username=username)
+        if len(user) == 0:
+            raise GitlabGetError(
+                "No users found when searching for username '%s'" % username, 404
+            )
+        return user[0]
```

Comment:
My understanding is that these functions are to support gitlabform config where we can refer to username or group name but the API expects their IDs. Please correct me if I'm mistaken.

I think we might have related code elsewhere because the same concept is supported in multiple places. For example: "allowed to merge", "allowed to push", etc in branch protection setting, or "allowed to create" in protect tags or "deploy access level" in protected environments. Could you check on that? Either we update that common code to python-gitlab library or use the existing library instead of adding new logic here.

What do you think?

Reply:
@TimKnight-DWP - I was just checking what the other processors are doing. It seems things aren't quite consistent already. For example:

The `tags_processor.py` has the logic to retrieve a user by username. This function is now duplicated in the this PR.

https://github.com/gitlabform/gitlabform/blob/6bd5fd2ba653f8e8cec759f7f1a94efff49c7b67/gitlabform/processors/project/tags_processor.py#L15-L21

The branch protection config processing also need to retrieve user id and group id:

https://github.com/gitlabform/gitlabform/blob/6bd5fd2ba653f8e8cec759f7f1a94efff49c7b67/gitlabform/processors/util/branch_protector.py#L142-L147

Looks like the common code for this is in `gitlabform/gitlab/core.py`:

https://github.com/gitlabform/gitlabform/blob/6bd5fd2ba653f8e8cec759f7f1a94efff49c7b67/gitlabform/gitlab/core.py#L82-L99

I have a suggestion but would love to hear your thoughts. I think when the entire project is converted to python-gitlab, we won't need `gitlabform/gitlab`. I'm wondering if we create a new utils in `gitlabform/processors/util/` (or another path that makes sense) that contains the common codes instead of where you have them now. We have only converted a few handful of processors to python-gitlab so far. By having that common code setup now, the remaining ones should re-use those where necessary. The only existing processor that might need another refactor I think is the `tags_processor`.

What do you think? If we do go with the suggestion above, maybe we should look into the `@functools.lru_cache()` annotation? It's being used for the existing code. I'm not experienced enough with this to say whether we need it or not but a [quick glance at the doc](https://docs.python.org/3/library/functools.html#functools.lru_cache) sounds like it might be useful.